File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import defaultdict

def solution(n, wires):
    nodes = defaultdict(list)
    for w in wires:
        nodes[w[0]].append(w[1])
        nodes[w[1]].append(w[0])
    
    def dfs(nodes, visited, node, count):
        visited.add(node)
        count[0] += 1
        for n in nodes[node]:
            if n not in visited:
                dfs(nodes, visited, n, count)
    
    result = float('inf')
    for w in wires:
        count = [0]
        nodes[w[0]].remove(w[1])
        nodes[w[1]].remove(w[0])
        visited = set()
        dfs(nodes, visited, w[0], count)
        b = n-count[0]
        result = min(result, abs(b-count[0]))
        nodes[w[0]].append(w[1])
        nodes[w[1]].append(w[0])
        count[0] = 0
    return result
        

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    Map<Integer, List<Integer>> map;

    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;

        for (int i = 0; i < wires.length; i++) {
            map = new HashMap<>();
            for (int k=0; k<n; k++) {
                map.put(k+1, new ArrayList<>());
            }

            int node = 0;

            for (int j=0; j<wires.length; j++) {
                if (i==j) node = i+1;
                if (i!=j) {
                    int[] wire = wires[j];
                    map.get(wire[0]).add(wire[1]);
                    map.get(wire[1]).add(wire[0]);
                }
            }

            answer = Math.min(answer, Math.abs(n - 2 * dfs(node)));
        }

        return answer;
    }

    public int dfs(int node) {
        Deque<Integer> stack = new ArrayDeque<>();
        boolean[] visited = new boolean[map.size()+1];
        stack.push(node);
        int cnt = 0;
        while(!stack.isEmpty()) {
            int n = stack.pop();
            cnt++;
            visited[n] = true;
            List<Integer> list = map.get(n);
            for (Integer i : list) {
                if (!visited[i]) stack.push(i);
            }
        }

        return cnt;
    }
}